### PR TITLE
Fix issue 2963

### PR DESCRIPTION
**Pull request by rahku**
PTAL @JohnChen0

### CODE REVIEW

**Comment by rahku** [comment]
Fixes #2963

**Comment by JohnChen0** [comment]
LGTM, though I'm not familiar with ARM64 architecture.

**Comment by rahku** [comment]
PTAL @JKOTAS

File: src/zap/zapcode.cpp [hunk 1]
```diff
@@ -1939,6 +1939,19 @@ DWORD ZapLazyHelperThunk::SaveWorker(ZapWriter * pZapWriter)
     if (pImage != NULL)
         pImage->WriteReloc(buffer, (int)(p - buffer), m_pTarget, 0, IMAGE_REL_BASED_THUMB_BRANCH24);
     p += 4;
+#elif defined(_TARGET_ARM64_)
+    // ldr x1, [PC+4]
+    *(DWORD *)(p) =0x58000021;
+    p += 4;
+    if (pImage != NULL)
+        pImage->WriteReloc(buffer, (int)(p - buffer), m_pArg, 0, IMAGE_REL_BASED_PTR);
```
Comment (jkotas) [inline src/zap/zapcode.cpp]: This pointer needs to be after the `b JIT_StrCns` - you cannot have a random pointer between two instructions.

File: src/zap/zaprelocs.cpp [hunk 2]
```diff
@@ -107,6 +107,20 @@ void ZapBaseRelocs::WriteReloc(PVOID pSrc, int offset, ZapNode * pTarget, int ta
...
 #endif
+#if defined(_TARGET_ARM64_)
+    case IMAGE_REL_ARM64_BRANCH26:
+        {
+            TADDR pSite = (TADDR)m_pImage->GetBaseAddress() + rva;
+
+            INT64 relOffset = (INT64)(pActualTarget - (pSite + sizeof(INT64)));
+            if (!FitsInRel28(relOffset))
+            {
+                ThrowHR(COR_E_OVERFLOW);
```
Comment (briansull) [inline src/zap/zaprelocs.cpp]: We might also need:
                  // Retry the compilation with IMAGE_REL_BASED_ARM64_BRANCH26 relocations disabled 
102                  // (See code:ZapInfo::getRelocTypeHint) 102                  // (See code:ZapInfo::getRelocTypeHint)

Comment (rahku) [inline src/zap/zaprelocs.cpp]: that is done for arm and only applies for desktop case and not coreclr. 
m_fNGenLastRetry is only valid for desktop and not for coreclr.
So not doing anything here for arm64

**Comment by rahku** [comment]
Updated PR. Can you look at the changes.

**Comment by briansull** [comment]
LGTM

**Comment by rahku** [comment]
@dotnet-bot test Windows_NT arm64

**Comment by jkotas** [comment]
LGTM. Squash before merging...

**Comment by bendono** [comment]
Hi @jkotas. Are you aware of the following feature? [Squash your commits](https://github.com/blog/2141-squash-your-commits)
Instead of asking everyone to squash their commits, the squash could be done at merge time.

**Comment by kyulee1** [comment]
@dotnet-bot test Windows_NT arm64

**Comment by jkotas** [comment]
@bendono Thanks for the tip - I will give it a try next time.

**Pull request merged**